Add iaf_psc_alpha_ps neuron model

Silmathoron opened the pull request:
This PR implements the ``iaf_psc_alpha_ps`` precise model.
Addresses #370 

I'll add a test ASAP.
For formatting, do we really still need to use ``clang-format-3.6`` ?! This doesn't exist anywhere anymore ^^

heplesser commented:
@Silmathoron We recently decided to name PRs so that the PR names directly turn into user-informative lines for the release notes. They should always start with a verb in present tense and should not contain references to issues. I therefore changed the title.

heplesser commented on precise/precisemodule.cpp:
```diff
@@ -48,6 +48,7 @@
 #include "model_manager_impl.h"
 
 // Includes from precise:
+#include "iaf_psc_alpha_ps.h"
```
Please put the includes in alphabetical order.

heplesser commented on precise/precisemodule.cpp:
```diff
@@ -91,6 +92,8 @@ void
 {
   // register models
+  kernel().model_manager.register_node_model< iaf_psc_alpha_ps >(
+    "iaf_psc_alpha_ps" );
```
Please put registrations in alphabetical order.

heplesser commented on pynest/nest/tests/test_iaf_psc_alpha_ps.py:
```diff
@@ -0,0 +1,190 @@
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+import unittest
+
```
I'd suggest to remove the empty lines between imports.

heplesser commented on pynest/nest/tests/test_iaf_psc_alpha_ps.py:
```diff
@@ -0,0 +1,190 @@
+Comparing the new implementations the precise ``iaf_psc_alpha_ps`` model to the
+previous reference implementation given by ``iaf_psc_alpha_canon``.
+
+The new implementation provides two separate buffer for excitatory and
```
buffers

heplesser commented on pynest/nest/tests/test_iaf_psc_alpha_ps.py:
```diff
@@ -0,0 +1,190 @@
+                rel_diff[model][record] = np.average(rds)
+        return rel_diff
+
+    def test_closeness_same_timescales(self):
```
time constants, not time scales, isn't it?

heplesser commented on pynest/nest/tests/test_iaf_psc_alpha_ps.py:
```diff
@@ -0,0 +1,190 @@
+        '''
+        Compare models for tau_syn_ex == tau_syn_in (should be equal).
+        '''
+        Ie = 370.
```
Insert empty line above.

Silmathoron replied:
You want an empty line between the docstring and ``Ie``?! Damn... I hate that... :'( is it a NEST convention?

heplesser replied:
I don't think we have an explicit convention, I just find it more legible. PEP257 requires an empty line behind class docstrings, but this is not the case here. Btw, PEP257 also wants double quotes instead of single quotes around docstrings ;)

heplesser commented on pynest/nest/tests/test_iaf_psc_alpha_ps.py:
```diff
@@ -0,0 +1,190 @@
+        '''
+        Check that the behavior differs if ``tau_syn_in != tau_syn_ex``
+        '''
+        Ie = 370.
```
Insert empty line above.

heplesser commented on pynest/nest/tests/test_iaf_psc_alpha_ps.py:
```diff
@@ -0,0 +1,190 @@
+        data_new = nest.GetStatus(mm_new, "events")[0]
+
+        # test before 500.
+        keep = data_new["times"] < 500.
```
keep seems a strange name. Could you add a comment on why you expect equality before 500, but not thereafter?

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -0,0 +1,548 @@
+
+/** @BeginDocumentation
+Name: iaf_psc_alpha_ps - Leaky integrate-and-fire neuron
+with alpha-shape postsynaptic currents (one timescale for excitatory and
```
I would remove the text in (), since this model should become the default. Also, in this model we rather have *different* timescales, as it is written now one could think there is just a single for ex *and* in.

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -0,0 +1,548 @@
+V_th         double - Spike threshold in mV.
+V_reset      double - Reset potential of the membrane in mV.
+tau_syn_ex   double - Rise time of the excitatory synaptic function in ms.
+tau_syn-in   double - Rise time of the inhibitory synaptic function in ms.
```
tau_syn_in

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -0,0 +1,548 @@
+    A general and efficient method for incorporating exact spike times in
+    globally time-driven simulations Front Neuroinformatics, 4:113
+
+Author: Tanguy Fardet (from Diesmann, Eppler, Morrison, Plesser, Straube)
```
from -> based on

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -0,0 +1,548 @@
+
+Receives: SpikeEvent, CurrentEvent, DataLoggingRequest
+
+SeeAlso: iaf_psc_alpha, iaf_psc_alpha_presc, iaf_psc_exp_ps
```
Why no reference to iaf_psc_alpha_canon?

Silmathoron replied:
Should I add it even though we're deprecating it?

heplesser replied:
Maybe not.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -0,0 +1,832 @@
+    V_m_ -= delta_EL;
+  }
+
+  updateValue< double >( d, names::I_syn_ex, I_ex_ );
```
The synaptic currents and their derivatives should not be settable. I don't think they are in any other models.

Silmathoron replied:
They were in [canon](https://github.com/nest/nest-simulator/blob/master/precise/iaf_psc_alpha_canon.cpp#L206) but this is set state, not parameters, so it seems normal to me

heplesser replied:
Your are right about `canon`, but I think that is rather a leftover from debugging. The derivatives are not documented in `canon`, and they cannot be set in the plain `iaf_psc_...` models. So I would argue against introducing them here.

jakobj commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -0,0 +1,832 @@
+  V_.psc_norm_ex_ = 1.0 * numerics::e / P_.tau_syn_ex_;
+  V_.psc_norm_in_ = 1.0 * numerics::e / P_.tau_syn_in_;
+
+  V_.gamma_ex_ = 1 / P_.c_m_ / ( 1 / P_.tau_syn_ex_ - 1 / P_.tau_m_ );
```
these do not seem numerically stable for `tau_syn_ex_/in_ == tau_m_`, neither do they seem to be in `iaf_psc_alpha_canon`. this should be fixed, but maybe not in this PR?

Silmathoron replied:
This is checked in the propagator (line 314 and below)

jakobj replied:
hmm, as far as I understand the propagators are selected according to whether `tau_syn_ex_ ~ tau_m_` and then computed such that division by zero is avoided. I am failing to see how the same is done for the gammas? maybe I need another coffee? ;)

Silmathoron replied:
No, obviously you're right! I see it now that I had mine ;)
I wouldn't mind if this was done in another PR, yes ^^"

jakobj replied:
ok, then I would suggest to throw an exception when the time constants are too close as a temporary solution. could you additionally please create an issue that describes what should be fixed here?

jakobj commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -0,0 +1,832 @@
+double
+nest::iaf_psc_alpha_ps::thresh_find1_( double const dt ) const
+{
+  double tau =
```
`const double`

jakobj commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -0,0 +1,832 @@
+
+// finds threshpassing
+inline double
+nest::iaf_psc_alpha_ps::thresh_find_( double const dt ) const
```
`double` <-> `const`?

jakobj commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -0,0 +1,832 @@
+
+// finds threshpassing via linear interpolation
+double
+nest::iaf_psc_alpha_ps::thresh_find1_( double const dt ) const
```
`double` <-> `const`?

jakobj commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -0,0 +1,832 @@
+
+// finds threshpassing via quadratic interpolation
+double
+nest::iaf_psc_alpha_ps::thresh_find2_( double const dt ) const
```
`double` <-> `const`?

jakobj commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -0,0 +1,832 @@
+}
+
+double
+nest::iaf_psc_alpha_ps::thresh_find3_( double const dt ) const
```
`double` <-> `const`?

Silmathoron commented:
@heplesser @jakobj I corrected spike handling (single buffer like the exp version) and incorporated most of your comments.

Regarding the tests, I removed ``canon`` wherever it seemed reasonable.
There are still a few ``iaf_ps_`` tests where I did not include the new model because ``iaf_psc_exp_ps`` was not included either and adding it would require removing the ``tau_syn`` variable which does not exist in the model. If you think it would be better, I can remove ``tau_syn`` from `` Set`` and include the model in the tests.
These tests are:
* ``test_iaf_ps_psp_accuracy.sli``
* ``test_iaf_ps_psp_poisson_generator_accuracy.sli``
* ``test_iaf_ps_psp_poisson_spike_accuracy.sli``

Silmathoron commented:
By the way, ``do_tests.py`` (called from ``make installcheck``) errors on current master because it tries to call ``nest.init``, which does not exist (``init`` is part of ``ll_api.py`` but is not imported to the main ``__init__.py``, so)

As far as I'm concerned, I think this is a bug and that ``nest.ll_api.init`` should be called instead. Could you confirm/infirm, and tell me if this is already known or if I should file a new issue?

lekshmideepu commented:
Please take a look [here](https://travis-ci.org/lekshmideepu/nest-simulator/jobs/521221141#L6860)
` File "/Users/travis/build/lekshmideepu/nest-simulator/result/share/nest/extras/do_tests.py", line 50, in <module>
    nest.init(["--verbosity=WARNING", "--quiet"])
AttributeError: 'module' object has no attribute 'init'
Assertion failed: (kernel_manager_instance_), function get_kernel_manager, file /Users/travis/build/lekshmideepu/nest-simulator/nestkernel/kernel_manager.h, line 206.
/Users/travis/build/lekshmideepu/nest-simulator/result/share/nest/extras/do_tests.sh: line 794: 21235 Abort trap: 6           "${PYTHON}" "${PYTHON_HARNESS}" >> "${TEST_LOGFILE}"
cat: pynest_test_numbers.log: No such file or directory
cat: pynest_test_numbers.log: No such file or directory
cat: pynest_test_numbers.log: No such file or directory
cat: pynest_test_numbers.log: No such file or directory`

jougs commented:
@Silmathoron: @lekshmideepu and I just found exactly the same problem with `do_tests.py` (see above) and also think that it should be `nest.ll_api.init(...)`. Alternatively we could make `init()` avaliable in `__all__`.

Everything works fine if you have `nosetests` installed.

lekshmideepu commented:
@Silmathoron : I would just fix it in my PR

Silmathoron commented:
@lekshmideepu which PR? Can you file an issue at the same time, then?
@jougs I don't think it belongs on the top level, I would just fix the call...

lekshmideepu commented:
@Silmathoron : sure I would file an issue

lekshmideepu commented:
#1181

heplesser commented on doc/guides/simulations_with_precise_spike_times.rst:
```diff
@@ -72,7 +72,7 @@ delta-shaped post-synaptic currents that employs precise spike times;
 precise location of an outgoing spike is determined analytically.
 
-``iaf_psc_alpha_canon`` and ``iaf_psc_alpha_presc`` are
+``iaf_psc_alpha_psc`` and ``iaf_psc_alpha_presc`` are
```
`..._ps`, not `..._psc`

heplesser commented:
The NEST Open Developer VC on 3 June approved the deprecation of `iaf_psc_alpha_canon` and its removal in NEST 3.0. 

The documentation of `iaf_psc_alpha_ps` should remark that this model replaced `iaf_psc_alpha_canon` from Morrison el al (2007).

suku248 commented:
In this model the threshold crossing is determined using interpolation as described in Morrison et al. (2007).
A more efficient way of locating the threshold crossing is bisectioning (Hanuschkin et al., 2010) as implemented in `iaf_psc_exp_ps`. I think the interpolation in `iaf_psc_alpha_ps` should also be replaced with bisectioning. If we don't replace it, `iaf_psc_exp_ps` and `iaf_psc_alpha_ps` have similar names but use different methods for finding the threshold crossing.
@heplesser what do you think?

heplesser commented:
@suku248 Thanks for spotting the difference in method between `_exp_ps` and this `_alpha_ps`.  I agree with you that we should be consistent across different models, so I would suggest adopting bisectioning also for `_alpha_ps`.

Silmathoron commented:
@heplesser @suku248 I updated the model with the bisectioning method, could you check?

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -45,15 +45,16 @@ namespace nest
 Name: iaf_psc_alpha_ps - Leaky integrate-and-fire neuron
-with alpha-shape postsynaptic currents.
+with alpha-shape postsynaptic currents and bisectioning method for
+approximation of threshold crossing..
```
Double fullstop at end of sentence.

Silmathoron replied:
```suggestion
approximation of threshold crossing.
```

heplesser commented on pynest/nest/lib/hl_api_helper.py:
```diff
@@ -69,7 +69,8 @@
                         'subnet': True,
-                        'aeif_cond_alpha_RK5': True}
+                        'aeif_cond_alpha_RK5': True,
+                        'iaf_psc_alpha_canon': True}
```
Also deprecated `iaf_psc_delta_canon`, it should be replaced with `iaf_psc_delta_ps` eventually.

heplesser replied:
Furthermore, deprecate `iaf_psc_alpha_presc`. This model will not be replace with anything else, it was just a temporary test model for the Morrison et al 2007 paper.

jougs commented:
@suku248: Are you OK with merging this? Doumo arigatou!
